cmd/oras/root/backup.go

```diff
@@ -0,0 +1,439 @@
+		if returnErr == nil {
+			returnErr = err
+		}
+	}()
+	if err := orasio.TarDirectory(tarFile, dstRoot); err != nil {
+		// remove the output file in case of error
+		if err := os.Remove(opts.output); err != nil && !errors.Is(err, fs.ErrNotExist) {
+			logger.Debugf("failed to remove output file %s: %v", opts.output, err)
+		}
+		return fmt.Errorf("failed to create tar archive at %s: %w", opts.output, err)
+	}
+	fi, err := os.Stat(opts.output)
+	if err != nil {
+		return fmt.Errorf("failed to stat output file %s: %w", opts.output, err)
+	}
+	return metadataHandler.OnTarExported(opts.output, fi.Size())
+}
+
+// resolveTags resolves tags to their descriptors.
+// It returns the resolved tags and their corresponding descriptors.
+func resolveTags(ctx context.Context, target oras.ReadOnlyTarget, specifiedTags []string) ([]string, []ocispec.Descriptor, error) {
+	var descs []ocispec.Descriptor
+	resolve := func(tags []string) error {
+		for _, tag := range tags {
+			desc, err := oras.Resolve(ctx, target, tag, oras.DefaultResolveOptions)
+			if err != nil {
+				return fmt.Errorf("failed to resolve tag %q: %w", tag, err)
+			}
+			descs = append(descs, desc)
+		}
+		return nil
+	}
+	if len(specifiedTags) > 0 {
+		// resolve the specified tags
+		descs = make([]ocispec.Descriptor, 0, len(specifiedTags))
+		if err := resolve(specifiedTags); err != nil {
+			return nil, nil, err
+		}
+		return specifiedTags, descs, nil
+	}
+
+	// discover all tags in the repository and resolve them
+	var tags []string
+	tagLister, ok := target.(registry.TagLister)
+	if !ok {
+		return nil, nil, errTagListNotSupported
+	}
+	if err := tagLister.Tags(ctx, "", func(gotTags []string) error {
+		if err := resolve(gotTags); err != nil {
+			return err
+		}
+		tags = append(tags, gotTags...)
+		return nil
+	}); err != nil {
+		return nil, nil, fmt.Errorf("failed to find tags: %w", err)
+	}
+	return tags, descs, nil
+}
+
+// parseArtifactReferences parses the input string into a repository
+// and a slice of tags.
+func parseArtifactReferences(artifactRefs string) (repository string, tags []string, err error) {
+	// validate input
+	if artifactRefs == "" {
+		return "", nil, errors.New("artifact reference cannot be empty")
+	}
+	// reject digest references early
+	if strings.ContainsRune(artifactRefs, '@') {
+		return "", nil, fmt.Errorf("digest references are not supported: %q", artifactRefs)
+	}
+	refParts := strings.Split(artifactRefs, ",")
+	mainRef := refParts[0]
+	extraTags := refParts[1:]
+
+	// validate repository
+	parsedRepo, err := registry.ParseReference(mainRef)
+	if err != nil {
+		return "", nil, fmt.Errorf("invalid reference %q: %w", mainRef, err)
+	}
+	mainTag := parsedRepo.Reference
+	parsedRepo.Reference = "" // clear the tag
+	repository = parsedRepo.String()
+
+	// validate each tag
+	for _, tag := range extraTags {
+		if tag == "" {
+			return "", nil, fmt.Errorf("empty tag in reference %q", artifactRefs)
+		}
+		if !tagRegexp.MatchString(tag) {
```

Comment:
Let's eliminate the use of `tagRegexp`.

```suggestion
		parsedRepo.Reference = tag
		if err := parsedRepo.ValidateReferenceAsTag(); err != nil {
```